plugins/imfile/imfile.c

```diff
@@ -2589,76 +2686,108 @@ fen_printevent(int event, char *pname)
-			"%s - errno %d\n", fobjp->fo_name, errno);
-		ABORT_FINALIZE(RS_RET_SYS_ERR);
-	} else {
-		/* Port successfull listening now*/
-		pLstn->bPortAssociated = 1;
+	pollFile(pLstn, NULL);	/* one final try to gather data */
+	lstnDel(pLstn);		/* Delete Listener now */
+	fileTableDelFile(&dirs[dirIdx].active, pLstn); /* Remove from filetable */
+	if(bDoRMState) {
+		DBGPRINTF("fen_removeFile unlinking '%s'\n", toDel);
+		if(unlink((char*)toDel) != 0) {
+			char errStr[1024];
+			rs_strerror_r(errno, errStr, sizeof(errStr));
+			errmsg.LogError(0, RS_RET_ERR, "fen_removeFile: could not remove state "
+				"file \"%s\": %s", toDel, errStr);
+		}
 	}
 finalize_it:
 	RETiRet;
+}
+
+static rsRetVal
+fen_removeDir(int dirIdx)
+{
+	DEFiRet;
 
+	DBGPRINTF("fen_removeDir removing dir (idx %d) '%s' \n", dirIdx, dirs[dirIdx].dirName);
+
+	/* Delete dir from dirs array! */
+	free(dirs[dirIdx].dirName);
+	if (dirs[dirIdx].dirNameBfWildCard != NULL)
+		free(dirs[dirIdx].dirNameBfWildCard);
+	free(dirs[dirIdx].active.listeners);
+	free(dirs[dirIdx].configured.listeners);
+	dirs[dirIdx].dirName = NULL;
+	dirs[dirIdx].dirNameBfWildCard = NULL;
+finalize_it:
+	RETiRet;
 }
-#endif
 
 static rsRetVal
-fen_processEvent(port_event_t* pportEvent, lstn_t *pLstn, int revents)
+fen_processEventFile(struct file_obj* fobjp, lstn_t *pLstn, int revents)
 {
-	struct file_obj *fobjp = (pportEvent == NULL ? &pLstn->pfinf->fobj : (struct file_obj*) pportEvent->portev_object);
-	int port = pLstn->pfinf->port;
 	struct stat statFile;
 	DEFiRet;
 
-	DBGPRINTF("imfile: fen_processEvent: %s (%s)\n", fobjp->fo_name, (pportEvent == NULL ? "NULL" : "EVENT"));
-
-	/* Port needs to be reassociated */
-	pLstn->bPortAssociated = 0;
+	// Use FileObj from listener if NULL
+	if (fobjp== NULL)
+		fobjp = &pLstn->pfinf->fobj;
 
+	/* uncomment if needed DBGPRINTF("fen_processEventFile: %s (0x%" PRIXPTR ") ", fobjp->fo_name, pLstn); **/
+	DBGPRINTF("fen_processEventFile: %s ", fobjp->fo_name);
 	if (revents) {
-		fen_printevent(revents, fobjp->fo_name);
+		fen_printevent(revents);
 	}
+	DBGPRINTF("\n");
+
+	if (pLstn == NULL) {
+		/* Should not be NULL but it case it is abort */
+		DBGPRINTF("fen_processEventFile: Listener '%s' for EVENT already deleted, aborting function.\n",
+			fobjp->fo_name);
+		FINALIZE;
+	}
+
+	/* Port needs to be reassociated */
+	pLstn->bPortAssociated = 0;
 
 	/* Compare filename first */
 	if (strcmp(fobjp->fo_name, (const char*)pLstn->pszFileName) == 0){
-		DBGPRINTF("imfile: fen_processEvent: matching file found: %s %s\n", fobjp->fo_name);
+		DBGPRINTF("fen_processEventFile: matching file found: '%s'\n", fobjp->fo_name);
 
 		/* Get File Stats */
 		if (!(revents & FILE_EXCEPTION) && stat(fobjp->fo_name, &statFile) == -1) {
-			DBGPRINTF("imfile: fen_processEvent: Failed to stat file "
-				": %s - errno %d\n", fobjp->fo_name, errno);
+			DBGPRINTF("fen_processEventFile: Failed to stat file: %s - errno %d\n",
```

Comment:
expose error to user?

Reply:
That's a good question. I would say not in this case. Because if user configured a logfile which has been created on rsyslog startup, rsyslog would output this error in each loop.

Reply:
ok - let's create an issue for this specifically. Probably requires wider audience feedback.